torch/_inductor/fx_passes/woq_mm.py

```diff
@@ -0,0 +1,115 @@
+import functools
+import logging
+
+import torch
+from ..._dynamo.utils import counters
+from ..pattern_matcher import fwd_only, joint_fwd_bwd, register_replacement
+
+log = logging.getLogger(__name__)
+aten = torch.ops.aten
+
+
+def _woq_mm_pattern_1(x, weight, scales):
+    return torch.nn.functional.linear(x, weight.to(dtype=x.dtype)) * scales
+
+
+def _woq_mm_replacement_1(x, weight, scales):
+    counters["inductor"]["woq_mm"] += 1
+    out_features = weight.size(0)
+    origin_x_size = x.size()
+    x = x.reshape(-1, origin_x_size[-1])
+    c = torch.ops.aten._weight_int8pack_mm(x, weight, scales)
+    new_shape = origin_x_size[:-1] + (out_features,)
+    return c.reshape(new_shape)
+
+
+def _woq_mm_params_check(match):
+    assert all(k in match.kwargs for k in ("x", "weight"))
+    x = match.kwargs["x"].meta["val"]
+    weight = match.kwargs["weight"].meta["val"]
+    # For now, we only support woq mm kernels
+    # with x.type=bfloat16 and w.type=int8
+    return (
+        x.dtype == torch.bfloat16
+        and weight.dtype == torch.int8
+        and x.device == weight.device
+    )
+
+
+def _woq_mm_extra_check(disable_cuda=False):
+    def fn(match):
+        if (
+            disable_cuda
+            and "x" in match.kwargs
+            and "cuda" in str(match.kwargs["x"].meta["val"].device)
+        ):
+            return False
+        return _woq_mm_params_check(match)
+
+    return fn
+
+
+def _get_woq_mm_patterns():
+    from .joint_graph import patterns
+
+    if torch.cuda.is_available():
+        # workaround https://github.com/pytorch/pytorch/issues/97894
+        device = "cuda"
+    else:
+        device = "cpu"
+
+    # sizes/values don't actually matter for initial trace
+    # once we get a possible match we re-trace with the actual values and verify the match still holds
+    x_inp = functools.partial(
+        torch.empty, (1, 1, 256), device=device, requires_grad=True
+    )
+    w_inp = functools.partial(torch.empty, (12, 256), device=device)
+    s_inp = functools.partial(torch.empty, (12), device=device)
+
+    x = functools.partial(x_inp, dtype=torch.bfloat16)
+    w = functools.partial(w_inp, dtype=torch.int8)
+    s = functools.partial(s_inp, dtype=torch.bfloat16)
+
+    workaround: dict[str, float]
+
+    for pattern, replacement, args, workaround, extra_check in [
```

Comment:
With one pattern that is (should be) inference only it's so not bad. But if we add more combinations here we need to aot serialize the patterns. can we add an assert that we only run this loop once with a TODO on serializing if we add more ?